doc/bgp_pic/bgp_pic.md

```diff
@@ -0,0 +1,200 @@
+	struct nexthop_group zd_ng;
+        /* PIC Nexthops */
+    uint32_t zd_pic_nhg_id;
+	struct nexthop_group zd_pic_ng;
+
+ These fields would be used in the following manner.
+
+| Cases |    Linux Kernel Update (slow path)   |      FPM (fast path)          |
+|:-----:|:------------------------------------:|:-----------------------------:|
+| No BGP PIC enabled | zd_ng is used as NHG  |    
+| BGP PIC enabled | zd_ng is used as NHG | zd_ng is used for PIC_CONTEXT, zd_pic_ng is used for NHG |
+
+### struct dplane_neigh_info
+This stucture is initialized via dplane_ctx_nexthop_init(), which is used to trigger NHG events. We don't need to make changes in this structure.
+
+## Zebra Modifications
+### BGP_PIC enable flag
+BGP_PIC_enable flag would be set based on zebra's command line arguments. This flag would be set only on the platform which Linux kernel supports NHG, a.k.a kernel_nexthops_supported() returns true.
+
+### Create pic_nhe
+From dplane_nexthop_add(), when normal NHG is created, we will try to create PIC NHG as well.
+zebra_nhe_find() is used to create or find a NHE. In create case, when NHE is for BGP PIC and BGP_PIC is enabled, we use the same same API to create a pic_nhe, a.ka. create nexthop with FORWARDING information only. The created pic_nhe would be stored in the new added field struct nhg_hash_entry *pic_nhe.
+
+### Handles kernel forwarding objects
+There is no change for zebra to handle kernel forwarding objects. Only zg_ng is used for NHG programming in kernel. 
+
+### Handles FPM forwarding objects
+#### Map Zebra objects to APP_DB via FPM
+When BGP_PIC is enabled, nhe's NHG would map to PIC_LIST, pic_nhe's NHG would map to forwarding NHG.
+Route object would use nhe's id as context id and use pic_nhe's id as NHG id.
+
+<figure align=center>
+    <img src="images/zebra_map_to_fpm_objs.jpg" >
+    <figcaption>Figure 4. Zebra maps forwarding objects to APP DB Objs when BGP PIC enables.<figcaption>
+</figure> 
+
+#### SRv6 VPN SAI Objects
+The following diagram shows SAI objects related to SRv6. The detail information could be found at
+https://github.com/opencomputeproject/SAI/blob/master/doc/SAI-IPv6-Segment-Routing-VPN.md
+
+<figure align=center>
+    <img src="images/srv6_sai_objs.png" >
+    <figcaption>Figure 5. SRv6 VPN SAI Objects<figcaption>
+</figure> 
+
+#### Map APP_DB to SAI objects
+<figure align=center>
+    <img src="images/app_db_to_sai.png" >
+    <figcaption>Figure 6. APP DB to SAI OBJs mapping<figcaption>
+</figure> 
+
+### Orchagent Modifications
+Handle two new forwarding objects from APP_DB, NEXTHOP_TABLE and PIC_CONTEXT_TABLE.  Orchagent would map proper objects to the proper SAI objects.
+
+## Zebra handles NHG member down events
+### Local link down events
+In Local link down event which would not triger BGP NH's reachability change case, we expect BGP NHG should not be updated. 
+TODO: Need to check with Kentaro to see if they would handle this event as the part of NHG handling.
```

Comment:
For your information, tests I made on an IGP setup with multipath with a BGP prefix 192.0.2.9/32 
BGP is informed of each nexthop update, and propagates the change by replacing the old routes.

case nexthop with 2 nexthops: shutdown on remote device from IGP.
```
# show bgp nexthop
[..]
 192.0.2.5 valid [IGP metric 30], #paths 1
  gate 172.31.0.3, if r1-eth1
  gate 172.31.2.4, if r1-eth2
  Last update: Sun Oct 15 20:52:09 2023
# shutdown 172.31.2.4 (not link down on if r1-eth2) : IGP updated and BGP nexthop is informed
# show bgp nexthop
 192.0.2.5 valid [IGP metric 30], #paths 1
  gate 172.31.0.3, if r1-eth1
  Last update: Sun Oct 15 20:59:54 2023

2023/10/15 20:59:55.082358 BGP: [ZJYCS-KQYD6] VRF default(0): Rcvd NH update 192.0.2.5/32(0)(0) - metric 30/30 #nhops 1/2 flags Valid Reg Label Valid 
2023/10/15 20:59:55.082375 BGP: [HJ3XJ-8EHF0]     nhop via 172.31.0.3 if 6 (1 labels)
2023/10/15 20:59:55.082380 BGP: [K7JBS-J1T3G] NH update for 192.0.2.5/32(0)(0)(VRF default) - flags Valid Reg Label Valid  chgflags Changed - evaluate paths
2023/10/15 20:59:55.082385 BGP: [JY1CY-BM14V] ... eval path 1/1 192.0.2.9/32 VRF default flags 0x418
[..]
2023/10/15 20:59:55.132473 BGP: [V15FP-4CPVK] Tx route add VRF 0 192.0.2.9/32 metric 0 tag 0 count 2 nhg 0
2023/10/15 20:59:55.132475 BGP: [J3BE9-CP44W]   nhop [1]: 192.0.2.5 if 0 VRF 0 wt 0  
```
case nexthop with 2 nexthops: shutdown on local device
An interface down event reaches BGP but does not trigger (this should be checked if not improved)
2 seconds later, BGP nexthop tracking is informed.
```
2023/10/15 21:04:57.234225 BGP: [N25MR-FXT2C] Rx Intf down VRF 0 IF r1-eth2
2023/10/15 21:04:57.234284 BGP: [ZN4WJ-AVQKV] Rx Intf address del VRF 0 IF r1-eth2 addr 172.31.2.1/24
2023/10/15 21:04:57.234297 BGP: [N25MR-FXT2C] Rx Intf down VRF 0 IF r1-eth2
2023/10/15 21:04:57.234404 BGP: [ZN4WJ-AVQKV] Rx Intf address del VRF 0 IF r1-eth2 addr fe80::305e:47ff:fed2:8752/64
2023/10/15 21:04:59.247552 BGP: [ZJYCS-KQYD6] VRF default(0): Rcvd NH update 192.0.2.5/32(0)(0) - metric 30/30 #nhops 1/2 flags Valid Reg Label Valid 
2023/10/15 21:04:59.247561 BGP: [HJ3XJ-8EHF0]     nhop via 172.31.0.3 if 6 (1 labels)
2023/10/15 21:04:59.247564 BGP: [K7JBS-J1T3G] NH update for 192.0.2.5/32(0)(0)(VRF default) - flags Valid Reg Label Valid  chgflags Changed - evaluate paths
2023/10/15 21:04:59.247569 BGP: [JY1CY-BM14V] ... eval path 1/1 192.0.2.9/32 VRF default flags 0x418
2023/10/15 21:04:59.297675 BGP: [V15FP-4CPVK] Tx route add VRF 0 192.0.2.9/32 metric 0 tag 0 count 2 nhg 0
2023/10/15 21:04:59.297677 BGP: [J3BE9-CP44W]   nhop [1]: 192.0.2.5 if 0 VRF 0 wt 0   
```

Reply:
For case 2, that is an optimization which we could add after enabling NHG in fpm. 
Currently, it seems although depend on NH resolution is set up, no one really uses it. That's why I ask Kentaro to see if they would pick it up or we will pick it up.

Reply:
For your case 1, it is explained below for IGP link failure handling, which could be local link failure or remote link failure.